src/lib.rs

```diff
@@ -1,21 +1,392 @@
+///
+/// ```
+/// extern crate starling;
+///
+/// # fn foo() -> starling::Result<()> {
+/// // Construct a new `Options` builder, providing the file containing
+/// // the main JavaScript task.
+/// starling::Options::new("path/to/main.js")
+///     // Finish configuring the `Options` builder and run the event
+///     // loop!
+///     .run()?;
+/// # Ok(())
+/// # }
+/// ```
+#[derive(Clone, Debug)]
+pub struct Options {
+    main: path::PathBuf,
+    sync_io_pool_threads: usize,
+    cpu_pool_threads: usize,
+    channel_buffer_size: usize,
 }
 
-#[test]
-fn it_works() {
+const DEFAULT_SYNC_IO_POOL_THREADS: usize = 8;
+const DEFAULT_CHANNEL_BUFFER_SIZE: usize = 4096;
+
+impl Options {
+    /// Construct a new `Options` object for configuring the `starling` event
+    /// loop.
+    ///
+    /// The given `main` JavaScript file will be evaluated as the main task.
+    pub fn new<P>(main: P) -> Options
+    where
+        P: Into<path::PathBuf>,
+    {
+        Options {
+            main: main.into(),
+            sync_io_pool_threads: DEFAULT_SYNC_IO_POOL_THREADS,
+            cpu_pool_threads: num_cpus::get(),
+            channel_buffer_size: DEFAULT_CHANNEL_BUFFER_SIZE,
+        }
+    }
+
+    /// Configure the number of threads to reserve for the synchronous IO pool.
+    ///
+    /// The synchronous IO pool is a collection of threads for shimming
+    /// synchronous IO libraries into the (otherwise completely asynchronous)
+    /// Starling system.
+    ///
+    /// ### Panics
+    ///
+    /// Panics if `threads` is 0.
+    pub fn sync_io_pool_threads(mut self, threads: usize) -> Self {
+        assert!(threads > 0);
+        self.sync_io_pool_threads = threads;
+        self
+    }
+
+    /// Configure the number of threads to reserve for the CPU pool.
+    ///
+    /// The CPU pool is a collection of worker threads for CPU-bound native Rust
+    /// tasks.
+    ///
+    /// Defaults to the number of logical CPUs on the machine.
+    ///
+    /// ### Panics
+    ///
+    /// Panics if `threads` is 0.
+    pub fn cpu_pool_threads(mut self, threads: usize) -> Self {
+        assert!(threads > 0);
+        self.cpu_pool_threads = threads;
+        self
+    }
+
+    /// Configure the size of mpsc buffers in the system.
+    ///
+    /// ### Panics
+    ///
+    /// Panics if `size` is 0.
+    pub fn channel_buffer_size(mut self, size: usize) -> Self {
+        assert!(size > 0);
+        self.channel_buffer_size = size;
+        self
+    }
+
+    /// Finish this `Options` builder and run the `starling` event loop with its
+    /// specified configuration.
+    pub fn run(self) -> Result<()> {
+        Starling::new(self)?.run()
+    }
+}
+
+impl Options {
+    // Get the number of `T`s that should be buffered in an mpsc channel for the
+    // current configuration.
+    fn buffer_capacity_for<T>(&self) -> usize {
+        let size_of_t = cmp::max(1, mem::size_of::<T>());
```

Comment:
Yes, any type that only has a single instance, or is only made up of zero sized things, is zero sized:

* `()`
* `struct Unit;`
* `struct LolYouThinkIHaveTInMeButIDont<T>(::std::marker::PhantomData<T>);`

Reply:
And those are still sendable and have an encoding size of 1?

Reply:
I'm not sure I understand what you mean by "sendable" and "encoding size" here.

The `Send` trait is orthogonal with the size of `Self`, and it doesn't involve any serialization/encoding.

Reply:
I guess what I'm asking is why it makes sense to say "things of size 0 need size 1 when buffered in an mpsc channel", as this code seems to do.

Reply:
First: to avoid a divide-by-zero error. Second: I think that the message must end up being wrapped in some non-zero sized type internally so that a channel can distinguish between an empty queue and a non-empty (but containing zero sized messages) queue.

Either way, I think this is perhaps getting ahead of ourselves a bit: we could probably just return the `options.channel_buffer_size` directly instead of trying to make all buffer sizes fit in a page...